File: internal/analyzers/common/plotpage/multipage.go
package plotpage

import (
	"fmt"
	"html/template"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	indexFileName    = "index.html"
	indexTitle       = "Analysis Report"
	indexDescription = "Select an analyzer to view its report."
)

// PageMeta carries metadata about a rendered analyzer page for the index.
type PageMeta struct {
	ID          string // Filename stem, e.g. "devs", "couples".
	Title       string // Display title, e.g. "Developer Contributions".
	Description string // Short description for the index card.
}

// MultiPageRenderer produces per-analyzer HTML pages plus an index page.
type MultiPageRenderer struct {
	OutputDir string // Directory to write HTML files into.
	Title     string // Project/report title shown on every page.
	Theme     Theme  // ThemeDark or ThemeLight.
}

// RenderAnalyzerPage renders a single analyzer page to <OutputDir>/<id>.html.
// Each page is standalone HTML with echarts + tailwind CDN and a navigation
// link back to index.html.
func (r *MultiPageRenderer) RenderAnalyzerPage(id, title string, sections []Section) error {
	page := NewPage(title, "")
	page.Theme = r.Theme
	page.ProjectName = r.Title

	navHTML, err := renderTemplate("nav.html", nil)
	if err != nil {
		return fmt.Errorf("render nav: %w", err)
	}

	renderer := HTMLRenderer{
		ExtraCSS: "",
	}

	// Prepend navigation as a section with no title (just the nav HTML).
	navSection := Section{
		Chart: rawHTML(navHTML),
	}
	page.Sections = append([]Section{navSection}, sections...)

	outPath := filepath.Join(r.OutputDir, id+".html")

	f, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("create %s: %w", outPath, err)
	}
	defer f.Close()

	renderErr := renderer.Render(f, page)
	if renderErr != nil {
		return fmt.Errorf("render %s: %w", id, renderErr)
	}

	return nil
}

// RenderIndex renders an index page with navigation cards to <OutputDir>/index.html.
func (r *MultiPageRenderer) RenderIndex(pages []PageMeta) error {
	page := NewPage(indexTitle, indexDescription)
	page.Theme = r.Theme
	page.ProjectName = r.Title

	indexContent, err := renderTemplate("index.html", indexData{Pages: pages})
	if err != nil {
		return fmt.Errorf("render index content: %w", err)
	}

	page.Sections = []Section{
		{Chart: rawHTML(indexContent)},
	}

	renderer := HTMLRenderer{}

	outPath := filepath.Join(r.OutputDir, indexFileName)

	f, err := os.Create(outPath)
	if err != nil {
		return fmt.Errorf("create %s: %w", outPath, err)
	}
	defer f.Close()

	renderErr := renderer.Render(f, page)
	if renderErr != nil {
		return fmt.Errorf("render index: %w", renderErr)
	}

	return nil
}

// RebuildIndex scans outputDir for *.html files (excluding index.html),
// derives page metadata from filenames, and regenerates index.html.
// This is useful after multiple renderers write pages to the same directory.
func (r *MultiPageRenderer) RebuildIndex() error {
	entries, err := os.ReadDir(r.OutputDir)
	if err != nil {
		return fmt.Errorf("read output dir: %w", err)
	}

	var pages []PageMeta

	for _, entry := range entries {
		name := entry.Name()
		if entry.IsDir() || !strings.HasSuffix(name, ".html") || name == indexFileName {
			continue
		}

		id := strings.TrimSuffix(name, ".html")
		title := strings.ReplaceAll(id, "-", "/")

		pages = append(pages, PageMeta{ID: id, Title: title})
	}

	sort.Slice(pages, func(i, j int) bool {
		return pages[i].Title < pages[j].Title
	})

	return r.RenderIndex(pages)
}

// indexData holds template data for index.html.
type indexData struct {
	Pages []PageMeta
}

// rawHTML is a Renderable that writes pre-rendered HTML.
type rawHTML template.HTML

// Render writes the raw HTML content.
func (r rawHTML) Render(w io.Writer) error {
	_, err := w.Write([]byte(r))
	if err != nil {
		return fmt.Errorf("write raw html: %w", err)
	}

	return nil
}

File: crates/cf-plotpage/src/multipage.rs
//! Multi-page renderer: per-analyzer HTML pages plus an index page with
//! navigation cards.

use std::fs;
use std::io;
use std::path::Path;

use crate::components::RawHtml;
use crate::page::{Page, Section};
use crate::templates::{render_index_content, NAV_HTML};
use crate::theme::Theme;

/// Index page constants.
const INDEX_FILE_NAME: &str = "index.html";
const INDEX_TITLE: &str = "Analysis Report";
const INDEX_DESCRIPTION: &str = "Select an analyzer to view its report.";

/// Metadata about a rendered analyzer page for the index.
#[derive(Debug, Clone, Default)]
pub struct PageMeta {
    /// Filename stem, e.g. `static-complexity`.
    pub id: String,
    /// Display title, e.g. `static/complexity`.
    pub title: String,
    /// Short description for the index card.
    pub description: String,
}

/// Produces per-analyzer HTML pages plus an index page.
pub struct MultiPageRenderer {
    /// Directory to write HTML files into.
    pub output_dir: String,
    /// Project/report title shown on every page.
    pub title: String,
    /// Color theme.
    pub theme: Theme,
}

impl MultiPageRenderer {
    /// Renders a single analyzer page to `<output_dir>/<id>.html`: standalone
    /// HTML with a navigation section prepended.
    ///
    /// # Errors
    /// Returns the I/O error when the output file cannot be written.
    pub fn render_analyzer_page(
        &self,
        id: &str,
        title: &str,
        sections: Vec<Section>,
    ) -> io::Result<()> {
        let mut page = Page::new(title, "");
        page.theme = self.theme;
        page.project_name.clone_from(&self.title);

        // Prepend navigation as a section with no title (just the nav HTML).
        let nav_section = Section {
            title: String::new(),
            subtitle: String::new(),
            hint: crate::page::Hint::default(),
            chart: Some(Box::new(RawHtml(NAV_HTML.to_string()))),
        };
        page.sections.push(nav_section);
        page.sections.extend(sections);

        let out_path = Path::new(&self.output_dir).join(format!("{id}.html"));
        fs::write(out_path, page.render())
    }

    /// Renders the index page with navigation cards to
    /// `<output_dir>/index.html`.
    ///
    /// # Errors
    /// Returns the I/O error when the output file cannot be written.
    pub fn render_index(&self, pages: &[PageMeta]) -> io::Result<()> {
        let mut page = Page::new(INDEX_TITLE, INDEX_DESCRIPTION);
        page.theme = self.theme;
        page.project_name.clone_from(&self.title);

        page.sections.push(Section {
            title: String::new(),
            subtitle: String::new(),
            hint: crate::page::Hint::default(),
            chart: Some(Box::new(RawHtml(render_index_content(pages)))),
        });

        let out_path = Path::new(&self.output_dir).join(INDEX_FILE_NAME);
        fs::write(out_path, page.render())
    }

    /// Scans the output dir for `*.html` files (excluding index.html), derives
    /// page metadata from filenames, and regenerates index.html.
    ///
    /// # Errors
    /// Returns the I/O error when the directory cannot be read or the index
    /// cannot be written.
    pub fn rebuild_index(&self) -> io::Result<()> {
        let mut pages: Vec<PageMeta> = Vec::new();
        for entry in fs::read_dir(&self.output_dir)? {
            let entry = entry?;
            let name = entry.file_name().to_string_lossy().into_owned();
            if entry.file_type()?.is_dir() || !name.ends_with(".html") || name == INDEX_FILE_NAME {
                continue;
            }
            let id = name.trim_end_matches(".html").to_string();
            let title = id.replace('-', "/");
            pages.push(PageMeta {
                id,
                title,
                description: String::new(),
            });
        }
        pages.sort_by(|a, b| a.title.cmp(&b.title));
        self.render_index(&pages)
    }
}
